### PR TITLE
Env

### PR DESCRIPTION
**Pull request by imrew-vantey**
Functions that can be used to change, set, unset and get from the environment.
And two functions that will take the environment and go from list to array and vise versa.

### CODE REVIEW

File: src/env/env.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+# include <libft.h>
+# include <stdlib.h>
+
+struct					s_env_list
+{
+	char				*key;
+	char				*value;
+	t_list_conn			conn;
+};
```
Comment (nloomans) [inline src/env/env.h]: I find the name `s_env_list` confusing, as this struct refers to a single element instead of the entire list. How about `s_env_pair`?

File: src/env/env.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+typedef struct			s_env
+{
+	t_list_meta			meta;
+}						t_env;
+
+struct s_env_list		*unpack_env(t_list_conn *const node);
```
Comment (nloomans) [inline src/env/env.h]: There are not needed externally. Could you move them to a `private.h`?

A `typedef struct s_env t_env;` will be needed so that the outside would can still reference `t_env`.

File: src/env/env.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+/*                                                 #+#    #+#                 */
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#ifndef ENV_H
+# define ENV_H
+
+# include <libft.h>
+# include <stdlib.h>
```
Comment (nloomans) [inline src/env/env.h]: The `<stdlib.h>` is not needed here, please include it in the `.c` files.

File: src/env/from_envp.c [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	{
+		new = (struct s_env_list *)ft_memalloc(sizeof(*new));
+		if (new == NULL)
+			return (-1);
+		assert(ft_strchr(envp[i], '=') != NULL);
+		equal_index = ft_strchr(envp[i], '=') - envp[i];
+		new->key = ft_strndup(envp[i], equal_index);
+		new->value = ft_strdup(&envp[i][equal_index + 1]);
+		if (new->key == NULL || new->value == NULL)
+			return (-1);
```
Comment (nloomans) [inline src/env/from_envp.c]: memory leak: `new` and either `new->key` or `new->value` it not freed here.

File: src/env/env.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+struct					s_env_list
+{
+	char				*key;
+	char				*value;
+	t_list_conn			conn;
+};
+
+typedef struct			s_env
+{
+	t_list_meta			meta;
```
Comment (nloomans) [inline src/env/env.h]: I find the name `meta` kinda vague here. How about `list`?

File: src/env/set.c [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+#include "env.h"
+
+int						env_set(t_env *const env,
+							char const *const key,
+							char const *const value)
+{
+	struct s_env_list	*new;
+
+	if (key == NULL || value == NULL)
+		return (-1);
```
Comment (nloomans) [inline src/env/set.c]: The `key` or `value` being `NULL` is a programming error and not something which should happen through normal execution. Either remove the `if` or replace it with an `assert`.

File: src/env/set.c [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	env_unset(env, key);
+	new = (struct s_env_list *)ft_memalloc(sizeof(*new));
+	if (new == NULL)
+		return (-1);
+	new->key = ft_strdup(key);
+	if (new->key == NULL)
+	{
+		free(new);
```
Comment (nloomans) [inline src/env/set.c]: Please try to use `ft_{mem,str}del` instead of `free`.

File: src/env/to_envp.c [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	ft_strdel(&temp);
+	if (env_str == NULL)
+		return (NULL);
+	temp = env_str;
+	env_str = ft_strjoin(temp, value);
+	ft_strdel(&temp);
+	if (env_str == NULL)
+		return (NULL);
+	return (env_str);
+}
```
Comment (nloomans) [inline src/env/to_envp.c]: This could be replaced with `ft_asprintf(&env_arr[i], "%s=%s", key, value);`

File: src/env/unpack.c [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+struct s_env_list			*unpack_env(t_list_conn *const node)
+{
+	if (node == NULL)
+	{
+		return (NULL);
+	}
+	else
+	{
+		return ((struct s_env_list *)\
```
Comment (nloomans) [inline src/env/unpack.c]: Unnecessary `\`

File: src/env/unset.c [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+	{
+		temp = unpack_env(*iter);
+		if (ft_strequ(temp->key, key) == 1)
+		{
+			del_holder = temp;
+			ft_list_unlink(&env->meta, &temp->conn);
+			ft_strdel(&del_holder->key);
+			ft_strdel(&del_holder->value);
+			ft_bzero(del_holder, sizeof(struct s_env_list));
+			free(del_holder);
```
Comment (nloomans) [inline src/env/unset.c]: The `ft_bzero` here doesn't really have an impact.

Also: Why doesn't `ft_list_unlink` set the `conn` pointers to NULL!? @nloomans please fix.

File: subprojects/libft/src/ft_strndup.c [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+	char	*dup;
+
+	if (len > ft_strlen(s1))
+		len = ft_strlen(s1);
+	dup = ft_strnew(len);
+	if (dup == NULL)
+		return (NULL);
+	ft_strlcpy(dup, s1, len + 1);
+	return (dup);
+}
```
Comment (nloomans) [inline subprojects/libft/src/ft_strndup.c]: Turns out I already have `ft_strsub`. Sorry about that

File: src/env/to_envp.c [hunk 12]
```diff
@@ -10,46 +10,25 @@
...
 	if (env_arr == NULL)
 		return (NULL);
-	while (i < env->meta.len && iter)
+	while (i < env->list.len && iter)
 	{
 		temp = unpack_env(iter);
-		env_arr[i] = fill_env_arr(temp->key, temp->value);
+		ft_asprintf(&env_arr[i], "%s=%s", temp->key, temp->value);
 		if (env_arr[i] == NULL)
 			return (NULL);
```
Comment (aholster) [inline src/env/to_envp.c]: free entire array incase of malloc failure.

File: src/env/from_envp.c [hunk 13]
```diff
@@ -11,27 +11,39 @@
...
 	size_t				i;
-	struct s_env_list	*new;
+	struct s_env_pair	*new;
 	size_t				equal_index;
 
 	i = 0;
 	while (envp[i] != '\0')
 	{
-		new = (struct s_env_list *)ft_memalloc(sizeof(*new));
+		new = (struct s_env_pair *)ft_memalloc(sizeof(*new));
```
[the remark this answers is not in the document]
Comment (nloomans) [reply src/env/from_envp.c]: [Objection!](https://i2.kym-cdn.com/entries/icons/facebook/000/004/590/OBJECTION!.jpg)

`sizeof(*new)` requires only the type of `new` to be updated for the malloc to work. While with `sizeof(struct s_env_pair)` you need to manually keep the type of `new` and the type inside of `sizeof` in sync.

File: src/env/from_envp.c [hunk 14]
```diff
@@ -13,13 +13,12 @@
...
 {
-	ft_memdel((void **)new);
-	if (new->key != NULL)
-		ft_strdel(&new->key);
-	if (new->value != NULL)
-		ft_strdel(&new->value);
+	ft_strdel(&new->key);
+	ft_strdel(&new->value);
+	ft_memdel((void **)&new);
+	return (-1);
```
Comment (aholster) [inline src/env/from_envp.c]: this faux return (ie always returning -1) due to a higher-level function having line-trouble is not acceptable.

File: src/env/from_envp.c [hunk 15]
```diff
@@ -11,27 +11,39 @@
...
 	size_t				equal_index;
 
 	i = 0;
 	while (envp[i] != '\0')
 	{
-		new = (struct s_env_list *)ft_memalloc(sizeof(*new));
+		new = (struct s_env_pair *)ft_memalloc(sizeof(*new));
 		if (new == NULL)
 			return (-1);
 		assert(ft_strchr(envp[i], '=') != NULL);
```
[the remark this answers is not in the document]
Comment (nloomans) [reply src/env/from_envp.c]: So I did some experimentation by creating two small programs. The parent program calls `execve` on the child program with some env variables that don't contain a `=`. The kernel didn't complain and the child program did get these env variables.

This means that getting a strange env like just the string `FOO` instead of `FOO=1` is possible. I'm not sure what to do in this case. I think just dropping in and possibly printing out a warning would be the most appropriate.

In any case the `assert` needs to go away since we now know that this is something which can happen.

[the remark this answers is not in the document]
Comment (imrew-vantey) [reply src/env/from_envp.c]: I removed the assert. Should I add anything?

File: src/env/from_envp.c [hunk 16]
```diff
@@ -44,7 +43,6 @@ int						env_from_envp(t_env *env, char **envp)
 			env_list_delete(env);
 			return (-1);
 		}
-		assert(ft_strchr(envp[i], '=') != NULL);
```
Comment (nloomans) [inline src/env/from_envp.c]: While the assert is gone, the code still assumes that a `=` is present. So it will now just crash when finding an env without a `=`.

Please change the code so it drops invalid environment variables and maybe even prints a warning if you feel like it. Some tests would also be nice.

**Comment by nloomans** [comment]
Got a compile error on Linux:

```
../src/env/from_envp.c: In function ‘env_from_envp’:
../src/env/from_envp.c:39:17: error: comparison between pointer and zero character constant [-Werror=pointer-compare]
   39 |  while (envp[i] != '\0')
      |                 ^~
../src/env/from_envp.c:39:9: note: did you mean to dereference the pointer?
   39 |  while (envp[i] != '\0')
      |         ^
cc1: all warnings being treated as errors
```

File: src/env/from_envp_test.c [hunk 17]
```diff
@@ -47,6 +47,29 @@ Test(env_from_envp, import_envp_containing_equal) {
...
+	cr_assert_eq(ret, 0);
+
+	value = env_get(&env, "FOO");
+	cr_assert_str_eq(value, "123");
+
+	value = env_get(&env, "UGH");
+	cr_assert_str_eq(value, "3");
+
+	value = env_get(&env, "BAR2345");
+	cr_assert_eq(value, NULL);
```
Comment (nloomans) [inline src/env/from_envp_test.c]: Right now the warning is printed out to the terminal when running tests:

```
Warning: env: 'BAR2345', no '=' present
[====] Synthesis: Tested: 22 | Passing: 22 | Failing: 0 | Crashing: 0
```

I would recommend redirecting the stderr and testing it's value so it doesn't get printed:

```diff
diff --git a/src/env/from_envp_test.c b/src/env/from_envp_test.c
index 72597ca..61ae672 100644
--- a/src/env/from_envp_test.c
+++ b/src/env/from_envp_test.c
@@ -11,6 +11,7 @@
 /* ************************************************************************** */

 #include <criterion/criterion.h>
+#include <criterion/redirect.h>
 #include "private.h"

 Test(env_from_envp, import_envp) {
@@ -47,7 +48,7 @@ Test(env_from_envp, import_envp_containing_equal) {
     cr_assert_str_eq(value, "ugh=1");
 }

-Test(env_from_envp, import_envp_no_equal) {
+Test(env_from_envp, import_envp_no_equal, .init = cr_redirect_stderr) {
     char        *arr[] = {"FOO=123", "BAR2345", "UGH=3", NULL};
     t_env       env;
     int         ret;
@@ -68,6 +69,8 @@ Test(env_from_envp, import_envp_no_equal) {

     value = env_get(&env, "BAR");
     cr_assert_eq(value, NULL);
+
+    cr_assert_stderr_eq_str("Warning: env: 'BAR2345', no '=' present\n");
 }

 Test(env_from_envp, import_envp_null) {
```

**Comment by nloomans** [comment]
@imrew-vantey there is still some white space at the end of lines left in `get_test.c`

File: src/env/get.c [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+/*                                                 #+#    #+#                 */
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "private.h"
+
+char					*env_get(t_env const *const env, char const *const key)
+{
+	struct s_env_pair	*temp;
```
Comment (aholster) [inline src/env/get.c:17]: this struct should probably be const

File: src/env/list_delete.c [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+/*       ivan-tey <ivan-tey@student.codam.nl>         +#+                     */
+/*       nloomans <nloomans@student.codam.nl>        +#+                      */
+/*                                                 #+#    #+#                 */
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "private.h"
+
+void					env_list_delete(t_env *env)
```
Comment (aholster) [inline src/env/list_delete.c:15]: since env itself is not a malloc'd object, the delete name is a misnomer, also, missing ptr const.

File: src/env/private.h [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+/*                                                 #+#    #+#                 */
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#ifndef PRIVATE_H
+# define PRIVATE_H
+
+# include "env.h"
+# include <libft.h>
```
Comment (aholster) [inline src/env/private.h:17]: switch order of headers

File: src/env/set.c [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+#include "private.h"
+#include <assert.h>
+
+int						env_set(t_env *const env,
+							char const *const key,
+							char const *const value)
+{
+	struct s_env_pair	*new;
+
+	assert(key != NULL || value != NULL);
```
Comment (aholster) [inline src/env/set.c:22]: assert is le wrong

File: src/env/set.c [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+/*   By: aholster <aholster@student.codam.nl>          +:+                    */
+/*       ivan-tey <ivan-tey@student.codam.nl>         +#+                     */
+/*       nloomans <nloomans@student.codam.nl>        +#+                      */
+/*                                                 #+#    #+#                 */
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "private.h"
+#include <assert.h>
```
Comment (aholster) [inline src/env/set.c:14]: switch header order

File: src/env/set.c (hunk 21 above)
Comment (aholster) [reply src/env/set.c:22]: oh also, is the sole assert so far, bit inconsistent

File: src/env/set.c [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+	env_unset(env, key);
+	new = ft_memalloc(sizeof(*new));
+	if (new == NULL)
+		return (-1);
+	new->key = ft_strdup(key);
+	if (new->key == NULL)
+	{
+		ft_memdel((void **)&new);
+		return (-1);
+	}
```
Comment (aholster) [inline src/env/set.c:32]: error check can be merged into the next one.

Comment (aholster) [reply src/env/set.c:32]: `	if (new->value == NULL || new->key == NULL)
	{
		ft_strdel(&new->key);
                ft_strdel(&new->value);
		ft_memdel((void **)&new);
		return (-1);
	}`

File: src/env/set_test.c [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+
+	ft_bzero(&env, sizeof(t_env));
+	ret = env_from_envp(&env, arr);
+	cr_assert_eq(ret, 0);
+
+	ret = env_set(&env, "UGH", "yasss");
+	cr_assert_eq(ret, 0);
+	value = env_get(&env, "UGH");
+	cr_assert_str_eq(value, "yasss");
+}
```
Comment (aholster) [inline src/env/set_test.c:78]: testing suite incomplete, missing test if rest of environment is mangled post-edit.
also missing test to check if env_set on empty environment results in correct behaviour

File: src/env/to_envp.c [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+/* ************************************************************************** */
+
+#include "private.h"
+#include "ft_printf.h"
+
+char					**env_to_envp(t_env *const env)
+{
+	size_t				i;
+	t_list_conn			*iter;
+	struct s_env_pair	*temp;
```
Comment (aholster) [inline src/env/to_envp.c:20]: make struct const

File: src/env/to_envp.c [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+		temp = unpack_env(iter);
+		ft_asprintf(&env_arr[i], "%s=%s", temp->key, temp->value);
+		if (env_arr[i] == NULL)
+		{
+			ft_arraydel((void ***)&env_arr, ft_memdel);
+			return (NULL);
+		}
+		iter = iter->next;
+		i++;
+	}
```
Comment (aholster) [inline src/env/to_envp.c:39]: add assert to assure `list.len` matches actual `len` rather than a double check as while condition

File: src/env/to_envp_test.c [hunk 27]
```diff
@@ -0,0 +1,31 @@
...
+	ft_bzero(&env_lst, sizeof(t_env));
+	ret = env_from_envp(&env_lst, arr);
+	cr_assert_eq(ret, 0);
+
+	env_arr = env_to_envp(&env_lst);
+	cr_assert_str_eq(env_arr[0], "FOO=1");
+	cr_assert_str_eq(env_arr[1], "BAR=2");
+	cr_assert_str_eq(env_arr[2], "UGH=3");
+	cr_assert_eq(env_arr[3], NULL);
+}
```
Comment (aholster) [inline src/env/to_envp_test.c:31]: Add additional cases:

1. env containing a variable such as `BAR=foo=nope`
2. empty env
3. variable such as `BAR=`

File: src/env/unset.c [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+/*       ivan-tey <ivan-tey@student.codam.nl>         +#+                     */
+/*       nloomans <nloomans@student.codam.nl>        +#+                      */
+/*                                                 #+#    #+#                 */
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "private.h"
+
+void					env_unset(t_env *env, char const *const key)
```
Comment (aholster) [inline src/env/unset.c:15]: missing const ptr for env

File: src/env/unset_test.c [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+	value = env_get(&env, "UGH");
+	cr_assert_eq(value, NULL);
+
+	value = env_get(&env, "BAR");
+	cr_assert_str_eq(value, "bar");
+	value = env_get(&env, "FOO");
+	cr_assert_str_eq(value, "foo");
+	value = env_get(&env, "NAHHH");
+	cr_assert_str_eq(value, "nahhh");
+}
```
Comment (aholster) [inline src/env/unset_test.c:106]: Add additional cases:

1. invoke unset on non-existent env entry
2. empty env
3. unset all?

File: src/env/env.h [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+int						env_from_envp(t_env *env, char **envp);
+char					**env_to_envp(t_env *const env);
+
+char					*env_get(t_env const *const env, char const *const key);
+void					env_unset(t_env *env, char const *const key);
+int						env_set(t_env *const env,
+								char const *const key,
+								char const *const value);
+
+void					env_list_delete(t_env *env);
```
[the remark this answers is not in the document]
Comment (nloomans) [reply src/env/env.h:27]: Could you rename it to `env_list_clear(t_env *const env);`?

File: src/env/env.h [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#ifndef ENV_H
+# define ENV_H
+
+typedef struct s_env	t_env;
+
+int						env_from_envp(t_env *env, char **envp);
```
[the remark this answers is not in the document]
Comment (nloomans) [reply src/env/env.h:18]: `env_from_envp(t_env *const env, char **const envp);` would be correct

File: src/env/from_envp.c [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+/*       nloomans <nloomans@student.codam.nl>        +#+                      */
+/*                                                 #+#    #+#                 */
+/*   License: GPLv3                                ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "private.h"
+#include "ft_printf.h"
+
+static int				fill_new(struct s_env_pair *new, char *envp)
```
[the remark this answers is not in the document]
Comment (nloomans) [reply src/env/from_envp.c:16]: `fill_new(struct s_env_pair *const new, const char *const envp)`

File: src/env/from_envp.c [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+	{
+		ft_strdel(&new->key);
+		ft_strdel(&new->value);
+		ft_memdel((void **)&new);
+		return (-1);
+	}
+	return (0);
+}
+
+int						env_from_envp(t_env *env, char **envp)
```
[the remark this answers is not in the document]
Comment (nloomans) [reply src/env/from_envp.c:38]: `env_from_envp(t_env *const env, char **const envp)`

File: src/env/get.c (hunk 18 above)
Comment (nloomans) [reply src/env/get.c:17]: `const struct s_env_pair	*temp;`

File: src/env/list_delete.c (hunk 19 above)
Comment (nloomans) [reply src/env/list_delete.c:15]: duplicate of previous comment, resolving

File: src/env/set.c (hunk 21 above)
Comment (nloomans) [reply src/env/set.c:22]: I would just remove the `assert` here, because it's not done anywhere else.

File: src/env/to_envp.c (hunk 25 above)
Comment (nloomans) [reply src/env/to_envp.c:20]: `const struct s_env_pair	*temp;`

File: src/env/to_envp.c (hunk 26 above)
Comment (nloomans) [reply src/env/to_envp.c:39]:
```
while (i < env->list.len)
{
    assert(iter != NULL);
    temp = unpack_env(iter);
```

File: src/env/unset.c (hunk 28 above)
Comment (nloomans) [reply src/env/unset.c:15]: `env_unset(t_env *const env, char const *const key)`

File: src/env/from_envp.c [hunk 34]
```diff
@@ -35,7 +35,7 @@ static int				fill_new(struct s_env_pair *new, char *envp)
 	return (0);
 }
 
-int						env_from_envp(t_env *env, char **envp)
+static int				fill_all(struct s_env *env, char **const envp)
```
Comment (aholster) [inline src/env/from_envp.c:38]: const pointer
